pkg/scheduler/plugins/coescheduling/core/core.go

```diff
@@ -0,0 +1,618 @@
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+	}
+	// check if gang is timeout by pod's Annotation
+	// todo: if not annotate, how can we handle the timeout event
+	if pod.Annotations[extension.AnnotationGangTimeout] == "true" {
+		return fmt.Errorf("gang has timeout, gangName: %v, podName: %v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+	}
+
+	if gang.getChildrenNum() < gang.getGangMinNum() {
+		return fmt.Errorf("gang child pod not collect enough, gangName: %v, podName: %v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+	}
+	// if gang is created by PodGroup, we check the cluster resource as Scheduler-Plugins does
+	if !PgFromAnnotation(pod) {
+		// TODO(cwdsuzhou): This resource check may not always pre-catch unschedulable pod group.
+		// It only tries to PreFilter resource constraints so even if a PodGroup passed here,
+		// it may not necessarily pass Filter due to other constraints such as affinity/taints.
+		nodes, err := pgMgr.snapshotSharedLister.NodeInfos().List()
+		if err != nil {
+			return err
+		}
+		pgFullName, pg := pgMgr.GetPodGroup(pod)
+		if pg == nil {
+			return nil
+		}
+		minResources := pg.Spec.MinResources.DeepCopy()
+		podQuantity := resource.NewQuantity(int64(pg.Spec.MinMember), resource.DecimalSI)
+		minResources[corev1.ResourcePods] = *podQuantity
+		err = CheckClusterResource(nodes, minResources, pgFullName)
+		if err != nil {
+			klog.ErrorS(err, "Failed to PreFilter", "podGroup", klog.KObj(pg))
+			return err
+		}
+	}
+
+	gangMode := gang.getGangMode()
+	if gangMode == extension.GangModeStrict {
+		podScheduleCycle := gang.getChildScheduleCycle(pod)
+		if !gang.isGangScheduleCycleValid() {
+			return fmt.Errorf("gang scheduleCycle not valid, gangName: %v, podName: %v",
+				gangId, GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		}
+		if podScheduleCycle >= gangScheduleCycle {
+			return fmt.Errorf("pod's schedule cycle too large, gangName: %v, podName: %v, podCycle: %v, gangCycle: %v",
+				gangId, GetNamespaceSplicingName(pod.Namespace, pod.Name), podScheduleCycle, gangScheduleCycle)
+		}
+	}
+	return nil
+}
+
+// Permit
+// we will calculate all Gangs in GangGroup whether the current number of assumed-pods in each Gang meets the Gang's minimum requirement.
+// and decide whether we should let the pod wait in Permit stage or let the whole gangGroup go binding
+func (pgMgr *PodGroupManager) Permit(ctx context.Context, pod *corev1.Pod) (time.Duration, Status) {
+	gangName := GetGangNameByPod(pod)
+	if gangName == "" {
+		return 0, PodGroupNotSpecified
+	}
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		gangName)
+	gang := pgMgr.cache.getGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return 0, PodGroupNotFound
+	}
+	// first add pod to the gang's WaitingPodsMap
+	gang.addAssumedPod(pod)
```

Comment:
I only find addAssumedPod, but has no delAssumedPod, if unreserve happened(especially in non-strict mode), should delAssumedPod

Reply:
yeah, delAssumedPod is in the podDeletedEvent handeld by the gangCache

Reply:
yes, delAssumedPod should happen during the Unreserve stage